contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,957 @@
+        // active trading position within the pool
+        if (_amountToSwap > 0) {
+            _swapToDesiredPosition(_amountToSwap, _minTokenReceived, _swapWeth);
+        }
+        // calling check liquidity early so we don't get unexpected errors when adding liquidity
+        // in the later stages of this function
+        if (tokenId != 0) {
+            _checkLiquidityWithinExpectedShare();
+        }
+
+        _addLiquidity();
+        _checkLiquidityWithinExpectedShare();
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    /**
+     * @dev Decrease partial liquidity from the pool.
+     * @param _partialLiquidityToDecrease The amount of liquidity to remove expressed in 18 decimals
+     */
+    function _removePartialLiquidity(uint256 _partialLiquidityToDecrease)
+        internal
+    {
+        require(_partialLiquidityToDecrease > 0, "Must remove some liquidity");
+        require(
+            _partialLiquidityToDecrease < 1e18,
+            "Mustn't remove all liquidity"
+        );
+
+        _removeLiquidity(_partialLiquidityToDecrease);
+    }
+
+    /**
+     * @dev Remove all liquidity from the pool.
+     */
+    // already check for reentrancy in withdrawAll
+    // slither-disable-start reentrancy-no-eth
+    function _removeAllLiquidity() internal {
+        _removeLiquidity(1e18);
+
+        positionManager.burn(tokenId);
+        tokenId = 0;
+        emit LiquidityTokenBurned(tokenId);
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    /**
+     * @dev Decrease partial or all liquidity from the pool.
+     * @param _liquidityToDecrease The amount of liquidity to remove expressed in 18 decimal point
+     */
+    function _removeLiquidity(uint256 _liquidityToDecrease) internal {
+        require(_liquidityToDecrease > 0, "Must remove some liquidity");
+
+        // unstake the position from the gauge
+        clGauge.withdraw(tokenId);
+
+        uint128 _liquidity = _getLiquidity();
+        // need to convert to uint256 since intermittent result is to big for uint128 to handle
+        uint128 liqudityToRemove = uint256(_liquidity)
+            .mulTruncate(_liquidityToDecrease)
+            .toUint128();
+
+        (uint256 _amountWeth, uint256 _amountOethb) = positionManager
+            .decreaseLiquidity(
+                // Both expected amounts can be 0 since we don't really care if any swaps
+                // happen just before the liquidity removal.
+                INonfungiblePositionManager.DecreaseLiquidityParams({
+                    tokenId: tokenId,
+                    liquidity: liqudityToRemove,
+                    amount0Min: 0,
+                    amount1Min: 0,
+                    deadline: block.timestamp
+                })
+            );
+
+        (
+            uint256 _amountWethCollected,
+            uint256 _amountOethbCollected
+        ) = positionManager.collect(
```

Comment:
Do we actually need to do this? This only collects fees, right? And aerodrome does no fess to LP users?

Reply:
Decrease liquidity doesn't transfer ERC20 tokens. After the decrease they are still on the pool contract but not in the LP position. 
That is why it is also annoying since there isn't a simple way to check if the pool has any LP deposited or not. Since just calling `balanceOf` with the pool's address also returns tokens that are owed as part of the liquidity decrease but not yet collected. 

https://github.com/aerodrome-finance/slipstream/blob/main/contracts/core/CLPool.sol#L519-L543

Reply:
👍